lib/resty/openssl/x509/store_ctx.lua

```diff
@@ -0,0 +1,234 @@
+
+local crl_stack_M = {}
+local STACK = "X509_CRL"
+local crl_stack_new = stack_lib.new_of(STACK)
+local crl_stack_add = stack_lib.add_of(STACK)
+local crl_stack_mt = stack_lib.mt_of(STACK, crl_lib.dup, crl_stack_M)
+
+_M.verify_flags = verify_flags
+
+-- to avoid circular reference among x509.store and x509.store_ctx,
+-- we do this check by ourselves
+local function istype_store(l)
+  return l and l.ctx and ffi.istype(x509_store_ptr_ct, l.ctx)
+end
+
+function _M.new(store, x509, untrusted_chain, properties)
+  local ctx
+
+  if store and not istype_store(store) then
+    return nil, "x509.store_ctx:new: expect a store instance at #1"
+  elseif x509 and not x509_lib.istype(x509) then
+    return nil, "x509.store_ctx:new: expect a x509 instance at #2"
+  elseif untrusted_chain and not chain_lib.istype(untrusted_chain) then
+    return nil, "x509.store_ctx:new: expect a x509.chain instance at #3"
+  end
+
+  if OPENSSL_3X then
+    ctx = C.X509_STORE_CTX_new_ex(ctx_lib.get_libctx(), properties)
+  else
+    ctx = C.X509_STORE_CTX_new()
+  end
+  if ctx == nil then
+    return nil, "x509.store_ctx.new: failed to create X509_STORE_CTX"
+  end
+
+  ffi_gc(ctx, C.X509_STORE_CTX_free)
+
+  local chain_dup_ctx
+  if untrusted_chain then
+    local chain_dup, err = chain_lib.dup(untrusted_chain.ctx)
+    if err then
+      return nil, err
+    end
+    chain_dup_ctx = chain_dup.ctx
+  end
+
+  local store_ctx
+  if store then
+    store_ctx = store.ctx
+  end
+
+  local x509_ctx
+  if x509 then
+    x509_ctx = x509.ctx
+  end
+
+  if C.X509_STORE_CTX_init(ctx, store_ctx, x509_ctx, chain_dup_ctx) ~= 1 then
+    return nil, format_error("x509.store_ctx:init: X509_STORE_CTX_init")
+  end
+
+  return setmetatable({
+    ctx = ctx,
+    verified = false,
+  }, mt), nil
+end
+
+function _M.istype(l)
+  return l and l.ctx and ffi.istype(store_ctx_ptr_ct, l.ctx)
+end
+
+function _M:set_default(verify_method)
+  if verify_method and C.X509_STORE_CTX_set_default(self.ctx, verify_method) ~= 1 then
+    return nil, "x509.store_ctx:set_default: invalid verify_method \"" .. verify_method .. "\""
+  end
+
+  return true
+end
+
+function _M:set_purpose(purpose)
+  if type(purpose) ~= "string" then
+    return nil, "x509.store:set_purpose: expect a string at #1"
+  end
+
+  local pchar = ffi.new("char[?]", #purpose, purpose)
+  local idx = C.X509_PURPOSE_get_by_sname(pchar)
+  idx = tonumber(idx)
+
+  if idx == -1 then
+    return nil, "invalid purpose \"" .. purpose .. "\""
+  end
+
+  local purp = C.X509_PURPOSE_get0(idx)
+  local i = C.X509_PURPOSE_get_id(purp)
+
+  if C.X509_STORE_CTX_set_purpose(self.ctx, i) ~= 1 then
+    return false, format_error("x509.store_ctx:set_purpose: X509_STORE_set_purpose")
+  end
+
+  return true
+end
+
+function _M:set_flags(...)
+  local flags = 0
+  for _, f in ipairs({...}) do
+    flags = bor(flags, f)
+  end
+
+  C.X509_STORE_CTX_set_flags(self.ctx, flags)
+
+  return true
+end
+
+function _M:add_crl(crl)
```

Comment:
i feel like this module `store_ctx` shares too much common API with `store` and may indicate we are not doing a proper API design. I would suggest the following:
- use `x509.store` as the only entrypoint when user tries to verify stuff.
- the `x509.store` instance creates a `X509_STORE` and `X509_STORE_CTX` under the hood
- every modified function like set_flags, add_crl goes to the `X509_STORE_CTX` ctx.
- add a new `x509.store:reset` function to drop the previous `X509_STORE_CTX`  ctx and create a new one, that disposes all previous flags, crl, etc.

Reply:
I would like to explain the initial motivation of adding `store_ctx`.

`store:verify` creates a temporary store ctx internally and does the verification. It's good for users to have this shortcut because in many cases the default configuration will be sufficient. But as we want to get more control, we add more and more parameters to `store:verify`. I don't feel it's a good way to making `store:verify` so many parameters. If we want to get more control of store ctx, creating a `store_ctx` and setting the ctx explicitly may be more flexible. We can extend other APIs easily and with the `store_ctx` reachable by users, they can also call the APIs that aren't supported by the module yet.

Another reason is we have the need for separate verification of revocation, but the current `store:verify` can't meet the need. If we keep the store ctx internal, then we need to pass many arguments to `check_revocation` as well and also lost the possibility of reusing the verified chain that is already built by previous calling of `store_ctx:verify`.

> the `x509.store` instance creates a `X509_STORE` and `X509_STORE_CTX` under the hood

This doesn't seem feasible. the `X509_STORE` and `X509_STORE_CTX` are not a 1-to-1 relationship. A store may have many store ctx at the same time.

> every modified function like set_flags, add_crl goes to the X509_STORE_CTX ctx.

I think the original idea of an additional level of `X509_STORE_CTX` over `X509_STORE` is putting the default settings into store and ctx inheriting and overriding the settings on demand. If we merge these two into one, then we will lose this ability.